### PR TITLE
Add ActivitySource.CreateActivity methods

### LINKED ISSUE #42784: Allow setting the Activity properties after Activity creation and before starting it when using ActivitySource
*The description and the Proposal edited by @tarekgh. The original description is moved down in the comment https://github.com/dotnet/runtime/issues/42784#issuecomment-777143962*

ActivitySource provides the method StartActivity which will try to create and start the Activity object. There is a scenario that needs to add some more properties to the Activity object after it is created and before it gets started. 
The reason is the caller wants to avoid creating the objects need to set to the Activity before ensuring the Activity object is created. It is important to have these properties set before starting the Activity so the listeners of the Start event will get the full set of the data.

# Proposal 

```C#
namespace System.Diagnostics
{
    public sealed class ActivitySource : IDisposable
    {
        // Alternative proposed name is CreateUnstartedActivity
        public Activity? CreateActivity(
                             string name,
                             ActivityKind kind, 
                             ActivityContext parentContext,
                             IEnumerable<KeyValuePair<string, object?>>? tags = null, 
                             IEnumerable<ActivityLink>? links = null);

        public Activity? CreateActivity(
                             string name, 
                             ActivityKind kind,
                             string parentContext,
                             IEnumerable<KeyValuePair<string, object?>>? tags = null, 
                             IEnumerable<ActivityLink>? links = null);
    }
}

```

# Example

```C#
    using (Activity activity = activitySource.CreateActivity("SomeName", ActivityKind.Server, context))
    {
        activity?.AddBaggage("db.type", "mongo");
        activity?.Start();

        // Do any necessary work      
    }
```

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes #42784

This PR should be addressing https://github.com/dotnet/runtime/issues/43853 too.

### CODE REVIEW

**Comment by tarekgh** [comment]
https://github.com/dotnet/runtime/issues/47313

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs [hunk 1]
```diff
@@ -1091,16 +1050,68 @@ public void SetCustomProperty(string propertyName, object? propertyValue)
...
+                activity.ActivityTraceFlags = parentContext.TraceFlags;
+                activity._traceState = parentContext.TraceState;
+            }
+
+            if (startTime != default)
+            {
+                activity.StartTimeUtc = startTime.UtcDateTime;
+            }
+
+            if (startIt)
```
Comment (noahfalk) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:1082]: Does anything prevent us from doing?
```suggestion
            if (startIt)
            {
                Start();
            }
```
It feels odd to be replicating all this start code.

File: src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs [hunk 2]
```diff
@@ -146,6 +146,9 @@ public sealed class ActivitySource : IDisposable
         public string Name { get { throw null; } }
         public string? Version { get { throw null; } }
         public bool HasListeners() { throw null; }
+        public System.Diagnostics.Activity? CreateActivity(string name, System.Diagnostics.ActivityKind kind) { throw null; }
+        public System.Diagnostics.Activity? CreateActivity(string name, System.Diagnostics.ActivityKind kind, System.Diagnostics.ActivityContext parentContext, System.Collections.Generic.IEnumerable<System.Collections.Generic.KeyValuePair<string, object?>>? tags = null, System.Collections.Generic.IEnumerable<System.Diagnostics.ActivityLink>? links = null) { throw null; }
+        public System.Diagnostics.Activity? CreateActivity(string name, System.Diagnostics.ActivityKind kind, string parentId, System.Collections.Generic.IEnumerable<System.Collections.Generic.KeyValuePair<string, object?>>? tags = null, System.Collections.Generic.IEnumerable<System.Diagnostics.ActivityLink>? links = null) { throw null; }
```
Comment (noahfalk) [inline src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs:151]: Weren't we going to handle #43853 by adding an IdFormat parameter here?

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs [hunk 3]
```diff
@@ -881,6 +881,98 @@ public void RestoreOriginalParentTest()
             }).Dispose();
         }
 
+
+        [ConditionalFact(typeof(RemoteExecutor), nameof(RemoteExecutor.IsSupported))]
+        public void TestActivityCreate()
```
Comment (noahfalk) [inline src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:886]: It would be nice to have at least one test case for each of the scenarios the API is intended to enable, ie 
(1) add tags/links/baggage after Create() and before Start()
(2) create a child Activity that has W3C format even when global format is Hierarchical

File: src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs (hunk 2 above)
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs:151]: No. After discussing this in the design review, @pakrym felt the other proposal is not needed and this can be handled by this CreateActivity. After creating the activity, they can set the IdFormat on such created Activity.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs (hunk 1 above)
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:1082]: I think it is possible. The only different case would be when someone calling StartActivity and passing ActivityContext with a default span Id. we used to keep `Activity.Parent=null` but if we call `Activity.Start()` it would set `Activity.Parent` to the Activity.Current. I am not sure though if this will be ok (or should be the right behavior). what you think?

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs (hunk 3 above)
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:886]: I'll add these tests

File: src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs (hunk 2 above)
Comment (noahfalk) [reply src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs:151]: Unforetunately that doesn't work. CreateActivity() calls the sampling callback, sampling callback has access to the ID, generating that ID has already fixed the id format. Using SetIdFormat() after CreateActivity() should probably throw an exception if it doesn't already.

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs:151]:
```C#
        public Activity SetIdFormat(ActivityIdFormat format)
        {
            if (_id != null || _spanId != null)
            {
                NotifyError(new InvalidOperationException(SR.SetFormatOnStartedActivity));
            }
            else
            {
                IdFormat = format;
            }
            return this;
        }
```


If the Activity not started, `_id` and `_spanId` would still be nulls. otherwise Activity.Start() would fail. no?

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs:151]: @noahfalk I have addressed all comments. The remaining point is this ne which related to IdFormat. I don't think we should fix the Ids before starting the Activity. Please let me know if you think otherwise.

Comment (noahfalk) [reply src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs:151]: I think we need to ensure that if we are creating a W3C activity then ActivityCreationOptions.TraceId that the sampler sees is the same trace Id that you get via Activity.TraceId. This implies it also shouldn't be legal to call SetIdFormat() after invoking the sampling callback because it would invalidate ActivityCreationOptions.TraceId that we already exposed.
It is probably easier to continue having _id and _spanId be computed lazily in Start() and that isn't a problem as long as we enforce the SetIdFormat() requirement some other way.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs (hunk 1 above)
Comment (noahfalk) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:1082]: In theory spanId = 0 is an invalid parent context. We should probably take a peak at the spec and chat with the OT folks to see if they have any input on what error handling behavior is desirable. I don't recall us ever defining the behavior very precisely but what I see the code currently do is give the bad id (in string or ActivityContext form) to the sampler as-is. Later in CreateAndStart we potentially ignore the id (if it was in string form) and create a new id based on IdFormat. This runs afoul of our sampler TraceId == Activity.TraceId invariant the same as above.

I don't care too much about the factoring on its own, but I am glad thinking about it sussed out issues like this one : )

File: src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs (hunk 2 above)
Comment (pakrym) [reply src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs:151]: >This implies it also shouldn't be legal to call SetIdFormat() after invoking the sampling callback because it would invalidate ActivityCreationOptions.TraceId that we already exposed.

This means we might need to bring the https://github.com/dotnet/runtime/issues/43853#issuecomment-780126996 feature request back.

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs:151]: Let me clarify the current behavior:

Currently if anyone calling StartActivity (which will be same as CreateActivity), 

- When using parent ActivityContext:
    - If the caller passing none-default ActivityContext (which has the traceId), then we honor that trace Id and that will never change when starting the Activity. i.e. what the listener see as trace id will be same as Activity traceId.
    - If the caller passing default ActivityContext, if there is valid parent (Activity.Current), we'll us the context of the parent and the trace Id again will be same for what the listener see and what Activity trace Id return.
    - If the caller passing default ActivityContext, if there is NO valid parent (Activity.Current == null)
        - If the listeners try to access the trace Id from the ActivityCreationOption, we'll create a new context including trace Id. and that trace Id will be used for Activity.TraceId too.
        - If the listeners didn't access the trace Id, we'll generate trace Id when starting the activity. **This is the only time IdFormat can affect generating the trace Id.**

- When using parent Id (string and not a context)
    - If the Id is formatted as a way can be converted to Activity Context, the generated context will be used as described before.
    - If the Id cannot be converted, we'll generate a new trace Id at the Activity.Start(). **This is the only time IdFormat will affect generating the trace Id.**

So to summarize, even when providing IdFormat parameter to StartActivity, it will not affect anything except the cases:
- The caller is passing default context, there is no parent (Activity.Current), and the listeners didn't try to access the trace Id from the provided ActivityCreationOptions.
- The caller is passing parent id string instead of the context and this parent Id is not formatted to extract context from it.

Considering these 2 cases, it can be achieved by CreateActivity too. 

In short, adding IdFormat to StartActivity wouldn't give us more than what CreateActivity will. And setting IdFormat wouldn't affect anything except the 2 mentioned cases.

So, I am still we don't need #43853 anyway. And CreateActivity still enough.

Comment (pakrym) [reply src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs:151]: >The caller is passing default context, there is no parent (Activity.Current), and the listeners didn't try to access the trace Id from the provided ActivityCreationOptions.

This case seems very real. Think of a worker app uploading a blob. There is no parent and all spans are exported so TraceId is not accessed.

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs:151]: > This case seems very real. Think of a worker app uploading a blob. There is no parent and all spans are exported so TraceId is not accessed.

Yes, agree. But using CreateActivity with the current design will make it work. right? I mean users can use CreateActivity and SetIdFormat if they need to for this scenario.

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs:151]: @pakrym I had offline discussion with @noahfalk and we propose adding extra boolean parameter to the CreateActivity called `forceW3CId`. The reason we didn't have the parameter of IdFormat because we want to promote W3C in general and not allow users to try the hierarchal one. Please let's know if you have any concern. I'll send it anyway to the fxdc and will CC you there.

```
        public Activity? CreateActivity(
                             string name,
                             ActivityKind kind, 
                             ActivityContext parentContext,
                             IEnumerable<KeyValuePair<string, object?>>? tags = null, 
                             IEnumerable<ActivityLink>? links = null,
                             bool forceW3CId = false);

        public Activity? CreateActivity(
                             string name, 
                             ActivityKind kind,
                             string parentContext,
                             bool forceW3CId = false,
                             IEnumerable<KeyValuePair<string, object?>>? tags = null, 
                             IEnumerable<ActivityLink>? links = null,
                             bool forceW3CId = false);
```

Comment (pakrym) [reply src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs:151]: LGTM. If we want to promote W3C can we default the parameter to true? 😄

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs:151]: > If we want to promote W3C can we default the parameter to true? 😄

I discussed that with @noahfalk and he preferred not to default the value to W3C as he want to have consistent behavior with StartActivity when passing the exact same parameters.

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs [hunk 4]
```diff
@@ -881,6 +889,311 @@ public void RestoreOriginalParentTest()
...
+        //  - If Activity.ForceDefaultIdFormat is true, then use the value Activity.DefaultIdFormat.  Otherwise, goto next step.
+        //  - If the parentId string is null and parent ActivityContext is default and we have Activity.Current != null, then use Activity.Current.IdFormat. Otherwise, goto next step.
+        //  - If we have non default parent ActivityContext, then use W3C. Otherwise, goto next step.
+        //  - if we have non null parent id string, try to parse it to W3C. If parsing succeeded then use W3C format. otherwise use Hierarchical format.
+        // --------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+
+        //           -----------------------------------------------------------------------------------------------------------------------------------------------------------------
+        //           Default Id Format, Id fromat to force, Parent id format,   parent id,  default parent context, validate trace Id,  expected create format, expected start format
```
Comment (noahfalk) [inline src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:1037]:
```suggestion
        //           Activity.DefaultIdFormat, Activity.Create(idFormat), Activity.Current.IdFormat,   parent id,  default parent context, validate trace Id,  expected create format, expected start format
        //           (ForceIdFormat=true)
```
I got the wrong idea about what these columns meant until I reverse engineered it from the source. This might help clarify the intent.

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs [hunk 5]
```diff
@@ -881,6 +889,311 @@ public void RestoreOriginalParentTest()
...
+                using ActivitySource aSource = new ActivitySource("FormatIdSource");
+                ActivityTraceId traceId = default;
+
+                //
+                //  Listener Creation
+                //
+
+                using ActivityListener listener = new ActivityListener();
+                listener.ShouldListenTo = (activitySource) => object.ReferenceEquals(aSource, activitySource);
+                listener.SampleUsingParentId = (ref ActivityCreationOptions<string> activityOptions) =>
```
Comment (noahfalk) [inline src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:1114]: Lets add a 2nd listener that only implements the Sample callback. This will let us confirm that Sample is always called for W3C formatted activities and never called for Hierarchal ones.

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs [hunk 6]
```diff
@@ -881,6 +889,311 @@ public void RestoreOriginalParentTest()
...
+        [InlineData($"W3C,              Hierarchical,       Unknown,            null,       true,                   false,              Hierarchical,           W3C")]
+        [InlineData($"Unknown,          Hierarchical,       W3C,                null,       true,                   false,              Hierarchical,           W3C")]
+        [InlineData($"W3C,              Hierarchical,       Hierarchical,       null,       true,                   false,              Hierarchical,           W3C")]
+        [InlineData($"W3C,              Hierarchical,       Hierarchical,       W3C,        true,                   false,              Hierarchical,           W3C")]
+        [InlineData($"Unknown,          Hierarchical,       Unknown,            null,       false,                  false,              Hierarchical,           W3C")]
+        [InlineData($"Unknown,          W3C,                Unknown,            null,       false,                  false,              W3C,                    W3C")]
+
+        [InlineData($"Unknown,          W3C,                Unknown,            null,       true,                   true,               W3C,                    Default")]
```
Comment (noahfalk) [inline src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:1070]: We should have the test case where a non-W3C string id gets passed as the parentId, the child does get W3C format, and the ActivityCreationOptions\<ActivityContext>.TraceId property does get viewed inside the Sample callback. That case wasn't previously calling Sample and we need to ensure that now it does work correctly.

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs [hunk 7]
```diff
@@ -881,6 +889,311 @@ public void RestoreOriginalParentTest()
...
+                //
+                //  Listener Creation
+                //
+
+                using ActivityListener listener = new ActivityListener();
+                listener.ShouldListenTo = (activitySource) => object.ReferenceEquals(aSource, activitySource);
+                listener.SampleUsingParentId = (ref ActivityCreationOptions<string> activityOptions) =>
+                {
+                    if (checkTraceId)
+                        traceId = activityOptions.TraceId;
```
Comment (noahfalk) [inline src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:1117]: Assuming we keep the previous behavior for SampleWithParentId, activityOptions.TraceId should always return default here.

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs [hunk 8]
```diff
@@ -881,6 +889,311 @@ public void RestoreOriginalParentTest()
...
+        [InlineData($"Unknown,          W3C,                Unknown,            null,       false,                  false,              W3C,                    W3C")]
+
+        [InlineData($"Unknown,          W3C,                Unknown,            null,       true,                   true,               W3C,                    Default")]
+        [InlineData($"W3C,              Unknown,            Unknown,            null,       true,                   true,               W3C,                    W3C")]
+        [InlineData($"Unknown,          Unknown,            W3C,                null,       true,                   true,               W3C,                    W3C")]
+        [InlineData($"Unknown,          Unknown,            Unknown,            W3C,        true,                   true,               W3C,                    W3C")]
+        [InlineData($"Unknown,          Unknown,            Unknown,            null,       false,                  true,               W3C,                    W3C")]
+
+        [ConditionalTheory(typeof(RemoteExecutor), nameof(RemoteExecutor.IsSupported))]
+        public void TestIdFormats(string data)
```
Comment (noahfalk) [inline src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:1077]: Nice test : )

**Review by noahfalk — COMMENTED** [review]
Mostly looked fine but I think we need to fix the closure allocation

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs [hunk 9]
```diff
@@ -135,8 +181,30 @@ public bool HasListeners()
...
             if (parentId != null)
             {
-                var aco = new ActivityCreationOptions<string>(this, name, parentId, kind, tags, links);
-                var acoContext = new ActivityCreationOptions<ActivityContext>(this, name, aco.GetContext(), kind, tags, links);
+                ActivityCreationOptions<string> aco = default;
+                ActivityCreationOptions<ActivityContext> acoContext = default;
+
+                // delegate to ensure updating the trace id in other ActivityCreationOption object when a new trace Id is generated.
+                // It is important so all listeners callbacks will always see the same trace Id which the activity will get created with
+                Action<ActivityTraceId, bool> updateTraceId = (traceId, isChangedFromContext) =>
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs:189]: Ok, I'll go with the #1 to make `ActivityCreationOptions<string>.TraceId` return default. I'll remove the delegate from the picture.

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs [hunk 10]
```diff
@@ -881,6 +889,311 @@ public void RestoreOriginalParentTest()
...
+        // Here is the sequence for how we determin the IdFormat of the Activity we create using ActivitySource:
+        // --------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+        //  - Use the IdFormat passed to the ActivitySource.Create if it is not "Unknown". Otherwise, goto next step.
+        //  - If Activity.ForceDefaultIdFormat is true, then use the value Activity.DefaultIdFormat.  Otherwise, goto next step.
+        //  - If the parentId string is null and parent ActivityContext is default and we have Activity.Current != null, then use Activity.Current.IdFormat. Otherwise, goto next step.
+        //  - If we have non default parent ActivityContext, then use W3C. Otherwise, goto next step.
+        //  - if we have non null parent id string, try to parse it to W3C. If parsing succeeded then use W3C format. otherwise use Hierarchical format.
+        // --------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:1034]: right. I handled this in the code :-) but didn't specify this step here. I'll add it here.

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs [hunk 11]
```diff
@@ -881,6 +889,311 @@ public void RestoreOriginalParentTest()
...
+        //           Default Id Format, Id fromat to force, Parent id format,   parent id,  default parent context, validate trace Id,  expected create format, expected start format
+        //           -----------------------------------------------------------------------------------------------------------------------------------------------------------------
+        [InlineData($"Unknown,          Unknown,            Unknown,            null,       true,                   false,              Default,                Default")]
+        [InlineData($"Unknown,          Unknown,            W3C,                null,       true,                   false,              W3C,                    W3C")]
+        [InlineData($"Unknown,          Unknown,            Hierarchical,       null,       true,                   false,              Hierarchical,           Hierarchical")]
+
+        [InlineData($"Unknown,          Unknown,            Unknown,            NonWC3Id,   true,                   false,              Hierarchical,           Hierarchical")]
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:1043]: will fix that.

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs (hunk 4 above)
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:1037]: will do

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs (hunk 6 above)
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:1070]: You mean when the listener is not providing `SampleUsingParentId`. right?

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs (hunk 7 above)
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:1117]: To clarify, it will return default only in the case the parent Id string cannot parsed to W3C. right?

**Comment by tarekgh** [comment]
@noahfalk I believe I have addressed all your feedback with the commit. please let me know if you have any more comments or we good to go. Thanks!

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs (hunk 7 above)
Comment (noahfalk) [reply src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:1120]: I was looking at the code more closely and may have misinterpretted the old behavior, trying to review more closely now : )

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs (hunk 6 above)
Comment (noahfalk) [reply src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs]: Right!

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs (hunk 7 above)
Comment (noahfalk) [reply src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:1120]: I did misinterpret and regardless I think the way you changed it will be fine. The behavior of ActivityCreationOptions\<string>.TraceId is a bit weird, but this matches it's previous weird behavior and breaking change isn't necessarily any better : )

**Review by noahfalk — APPROVED** [review]
LGTM, thanks Tarek!
